token_with_payload/src/contract.rs

```diff
@@ -0,0 +1,290 @@
+use crate::admin::{check_admin, has_administrator, write_administrator};
+use crate::allowance::{read_allowance, spend_allowance, write_allowance};
+use crate::auth::{PayloadTrait, SignaturePayload, SignaturePayloadV0};
+use crate::balance::{read_balance, receive_balance, spend_balance};
+use crate::balance::{read_state, write_state};
+use crate::metadata::{
+    read_decimal, read_name, read_symbol, write_decimal, write_name, write_symbol,
+};
+use crate::storage_types::DataKey;
+use soroban_auth::check_auth;
+use soroban_auth::{Identifier, Signature};
+use soroban_sdk::{contractimpl, symbol, BigInt, Bytes, Env, IntoVal};
+
+use soroban_sdk::{BytesN, Map, RawVal, Symbol, Vec};
+
+pub trait TokenTrait {
+    fn initialize(e: Env, admin: Identifier, decimal: u32, name: Bytes, symbol: Bytes);
+
+    fn nonce(e: Env, id: Identifier) -> BigInt;
+
+    fn allowance(e: Env, from: Identifier, spender: Identifier) -> BigInt;
+
+    fn approve(
+        e: Env,
+        from: Identifier,
+        nonce: BigInt,
+        spender: Identifier,
+        amount: BigInt,
+        sigs: Map<Identifier, Signature>,
+    );
+
+    fn balance(e: Env, id: Identifier) -> BigInt;
+
+    fn is_frozen(e: Env, id: Identifier) -> bool;
+
+    fn xfer(e: Env, from: Signature, nonce: BigInt, to: Identifier, amount: BigInt);
+
+    fn xfer_from(
+        e: Env,
+        spender: Signature,
+        nonce: BigInt,
+        from: Identifier,
+        to: Identifier,
+        amount: BigInt,
+    );
+
+    fn burn(e: Env, admin: Signature, nonce: BigInt, from: Identifier, amount: BigInt);
+
+    fn freeze(e: Env, admin: Signature, nonce: BigInt, id: Identifier);
+
+    fn mint(e: Env, admin: Signature, nonce: BigInt, to: Identifier, amount: BigInt);
+
+    fn set_admin(e: Env, admin: Signature, nonce: BigInt, new_admin: Identifier);
+
+    fn unfreeze(e: Env, admin: Signature, nonce: BigInt, id: Identifier);
+
+    fn decimals(e: Env) -> u32;
+
+    fn name(e: Env) -> Bytes;
+
+    fn symbol(e: Env) -> Bytes;
+}
+
+fn read_nonce(e: &Env, id: &Identifier) -> BigInt {
+    let key = DataKey::Nonce(id.clone());
+    e.contract_data()
+        .get(key)
+        .unwrap_or_else(|| Ok(BigInt::zero(e)))
+        .unwrap()
+}
+
+fn verify_and_consume_nonce(e: &Env, id: &Identifier, expected_nonce: &BigInt) {
+    match id {
+        Identifier::Contract(_) => {
+            if BigInt::zero(&e) != expected_nonce {
+                panic!("nonce should be zero for Contract")
+            }
+            return;
+        }
+        _ => {}
+    }
+
+    let key = DataKey::Nonce(id.clone());
+    let nonce = read_nonce(e, id);
+
+    if nonce != expected_nonce {
+        panic!("incorrect nonce")
+    }
+    e.contract_data().set(key, &nonce + 1);
+}
+
+pub struct Token;
+
+#[cfg_attr(feature = "export", contractimpl)]
+#[cfg_attr(not(feature = "export"), contractimpl(export = false))]
```

Comment:
Just an fyi, this doesn't do anything anymore. `contractimpl` always exports.

Reply:
Good to know. I copied and pasted this from the token example, so it should be removed there too.